src/devices/src/virtio/vsock/device.rs

```diff
@@ -45,54 +52,155 @@ use super::{defs, defs::uapi, EpollConfig};
-        for _ in defs::QUEUE_SIZES {
-            queue_evts.push(EventFd::new(libc::EFD_NONBLOCK).map_err(VsockError::EventFd)?);
-        }
-
-        let queues = defs::QUEUE_SIZES
+    pub fn new(cid: u64, mem: GuestMemory, backend: B) -> super::Result<Vsock<B>> {
+        let queues: Vec<VirtQueue> = defs::QUEUE_SIZES
             .iter()
-            .map(|&s| VirtQueue::new(s))
+            .map(|&max_size| VirtQueue::new(max_size))
             .collect();
+        let mut queue_events = Vec::new();
+        for _ in 0..queues.len() {
+            queue_events.push(EventFd::new(libc::EFD_NONBLOCK).map_err(VsockError::EventFd)?);
+        }
 
         Ok(Vsock {
             cid,
+            queues,
+            queue_events,
+            mem,
+            backend,
             avail_features: AVAIL_FEATURES,
             acked_features: 0,
-            epoll_config,
-            backend: Some(backend),
             interrupt_status: Arc::new(AtomicUsize::new(0)),
             interrupt_evt: EventFd::new(libc::EFD_NONBLOCK).map_err(VsockError::EventFd)?,
-            queues,
-            queue_evts,
+            activate_evt: EventFd::new(libc::EFD_NONBLOCK).map_err(VsockError::EventFd)?,
             device_activated: false,
         })
     }
+
+    pub fn cid(&self) -> u64 {
+        self.cid
+    }
+
+    /// Signal the guest driver that we've used some virtio buffers that it had previously made
+    /// available.
+    pub fn signal_used_queue(&self) -> result::Result<(), DeviceError> {
+        debug!("vsock: raising IRQ");
+        self.interrupt_status
+            .fetch_or(VIRTIO_MMIO_INT_VRING as usize, Ordering::SeqCst);
+        self.interrupt_evt.write(1).map_err(|e| {
+            error!("Failed to signal used queue: {:?}", e);
+            DeviceError::FailedSignalingUsedQueue(e)
+        })
+    }
+
+    /// Walk the driver-provided RX queue buffers and attempt to fill them up with any data that we
+    /// have pending.
+    pub fn process_rx(&mut self) -> bool {
+        debug!("vsock: epoll_handler::process_rx()");
+
+        let mut have_used = false;
+
+        while let Some(head) = self.queues[RXQ_INDEX].pop(&self.mem) {
+            let used_len = match VsockPacket::from_rx_virtq_head(&head) {
+                Ok(mut pkt) => {
+                    if self.backend.recv_pkt(&mut pkt).is_ok() {
+                        pkt.hdr().len() as u32 + pkt.len()
+                    } else {
+                        // We are using a consuming iterator over the virtio buffers, so, if we can't
+                        // fill in this buffer, we'll need to undo the last iterator step.
+                        self.queues[RXQ_INDEX].undo_pop();
+                        break;
+                    }
+                }
+                Err(e) => {
+                    warn!("vsock: RX queue error: {:?}", e);
+                    0
+                }
+            };
+
+            have_used = true;
```

Comment:
Shouldn't this be set to true only when `recv_pkt` is ok? Also, can we find a better name?

Reply:
When `recv_pkt` is not ok, the `break` on line 141 takes us out of the `while` loop and this no longer executes. Cannot think of a better one right now, but happy to take any suggestions :-s